proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,399 @@
+Safe Fixed Size Buffers
+=====
+
+## Summary
+
+Provide a general-purpose and safe mechanism for declaring fixed sized buffers within C# classes, structs, and interfaces.
+
+## Motivation
+
+This proposal plans to address the many limitations of https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/unsafe-code.md#228-fixed-size-buffers.
+Specifically it aims to allow the declaration of safe `fixed` buffers for managed and unmanaged types in a `struct`, `class`, or `interface`, and provide language safety verification for them.
+
+## Detailed Design (Option 1)
+
+The grammar for *variable_declarator* in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/classes.md#145-fields
+will be extended to allow specifying the size of the buffer:
+
+``` antlr
+field_declaration
+    : attributes? field_modifier* type variable_declarators ';'
+    ;
+
+field_modifier
+    : 'new'
+    | 'public'
+    | 'protected'
+    | 'internal'
+    | 'private'
+    | 'static'
+    | 'readonly'
+    | 'volatile'
+    | unsafe_modifier   // unsafe code support
+    ;
+
+variable_declarators
+    : variable_declarator (',' variable_declarator)*
+    ;
+    
+variable_declarator
+    : identifier ('=' variable_initializer)?
+    | fixed_size_buffer_declarator
+    ;
+    
+fixed_size_buffer_declarator
+    : identifier '[' constant_expression ']'
+    ;    
+```
+
+A *fixed_size_buffer_declarator* introduces a fixed-size buffer of a given element type.
+
+The buffer element type is the *type* specified in `field_declaration`. A fixed-size buffer declarator introduces a new member and consists of an identifier that names the member, followed by a constant expression enclosed in `[` and `]` tokens. The constant expression denotes the number of elements in the member introduced by that fixed-size buffer declarator. The type of the constant expression must be implicitly convertible to type `int`, and the value must be a non-zero positive integer.
+
+The elements of a fixed-size buffer shall be laid out sequentially in memory as though they are elements of an array.
+
+A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.
+
+Depending on the situation (details are specified below), an access to a fixed-size buffer member is classified as a value (never a variable) of either
+`System.ReadOnlySpan<S>` or `System.Span<S>`, where S is the element type of the fixed-size buffer. Both types provide indexers returning a reference to a
+specific element with appropriate "readonly-ness", which prevents direct assignment to the elements when language rules don't permit that.
+
+This limits the set of types that can be used as a fixed-size buffer element type to types that can be used as type arguments. For example, a pointer type cannot be used as an element type.
+
+The resulting span instance will have a length equal to the size declared on the fixed-size buffer.
```

Comment:
> The resulting span instance will have a length equal to the size declared on the fixed-size buffer.

It seems this might not be always true if the Span is created by user-supplied buffer type (from metadata).

Reply:
> It seems this might not be always true if the Span is created by user-supplied buffer type (from metadata).

I am not sure why that would make the difference. The attribute emitted on the type is supposed to match the size. Malformed metadata isn't something that the language spec is concerned about.